### ISSUE TITLE
Make dealing with large `ScVal` integers easier.

### ISSUE THREAD

**Issue by Shaptic** [stellar/js-stellar-base#610]
**Is your feature request related to a problem? Please describe.**
Dealing with `xdr.Int128Parts`, `xdr.Hyper`, etc. is a pain in the ass. Nobody outside of this library should have to deal with it.

**Describe the solution you'd like**
Abstractions around these values. I'd like something like:

```ts
type int64_t = number | bigint | string | xdr.Int64;
class I128 {
  constructor(lo: int128_t, hi?: int128_t);
  toNumber(): number; // throws if out of range
  toBigInt(): bigint;
  toHyper(): xdr.Hyper;
  toUnsignedHyper(): xdr.UnsignedHyper;
  toI128(): ScVal; // with ScValType = I128
  toU128(): ScVal; // with ScValType = U128
}
```

and support for this value when building transactions / operations / contract invocations.

**Describe alternatives you've considered**
Pain & suffering.

**Additional context**
A ton of work has already been done by @orbitlens here: stellar/js-xdr#96 as a starting point.

**Comment by Shaptic** [comment]
@piyalbasu @aristidesstaffieri @paulbellamy @sreuland, I've drafted a proposal in the [`bignumber-wrapper`](https://github.com/stellar/js-stellar-base/tree/bignumber-wrapper) branch that combines @orbitlens' work in https://github.com/stellar/js-xdr/pull/96 with a high-level wrapper oriented towards `ScVal`s. The details are in [`src/numbers/generic.js`](https://github.com/stellar/js-stellar-base/blob/bignumber-wrapper/src/numbers/generic.js) (and you can see more usage in [`test/unit/i256_test.js`](https://github.com/stellar/js-stellar-base/blob/bignumber-wrapper/test/unit/i256_test.js)), but I've also replicated the jsdoc example below for ease of reading on how this can be used by downstream systems:

```js
import * as sdk from "stellar-base";

// You have an ScVal from a contract and want to parse it into JS native.
const value = sdk.xdr.ScVal.fromXDR(someXdr, "base64");
let bigi = sdk.BigInteger.fromScVal(value);

// You have a number and want to shove it into a contract.
bigi = sdk.BigInteger([1, 2, 3, 4]);
bigi.toBigInt() // returns BigInt((4 << 96) | (3 << 64) | (2 << 32) | 1)
bigi.toNumber() // will throw: too large

// Pass any to e.g. a Contract.call(), conversion happens automatically.
const scValU128 = bigi.toU128();
const scValI256 = bigi.toI256();
const scValU64  = bigi.toU64();

// Lots of ways to initialize:
sdk.BigInteger("123456789123456789")
sdk.BigInteger(123456789123456789n);
sdk.BigInteger(["1234", "5678"]);
sdk.BigInteger([1234, "5678", 9012n, -1]);

// If you are confident in what you're doing and want to access `.raw`
// directly (which is faster than conversions), you can specify the type
// (otherwise, it's interpreted from the numbers you pass in):
const i = sdk.BigInteger(
  [1234, 5678, 9012, 3456], 
  { type: "u256" }
);
i.raw;            // an sdk.U256, which can be converted to an ScVal directly
i.raw.toScVal();  // an xdr.ScVal with ScValType = "scvU256"
```

Any and all feedback on this is appreciated! 

What other methods / helpers / initialization methods would you like to see?

**Comment by piyalbasu** [comment]
This is looking great, just had a couple of thoughts:

1. Should this include a .toU32 method to convert JS numbers to u32's, as well? I haven't tried yet, but I imagine we would need to do something similar

2. Once we're able to land this, it'd be great to have some string helper methods, as well. We ran into some issues trying to convert a string to Symbols, for ex, when initializing a token contract

**Comment by sreuland** [comment]
still absorbing it, but can see already how it would replace https://github.com/stellar/soroban-example-dapp/blob/main/convert.ts which is really nice.

**Comment by orbitlens** [comment]
A few notes: 
- `BigInteger` is not the best naming choice because it resembles JS native `BigInt`. This may cause a confusion. `ScInt`, `BigIntWrapper`, `GenericInt`, or something similar might be a better option.
- Automatic int size detection using `nearestInt()` method will cause problems. Need to provide the int type explicitly in the constructor (preferably, as the first argument), otherwise there always will be confusion between signed/unsigned and size. E.g. `sdk.BigInteger([1n, 1n])` will always return an `i64` value if I don't provide the type. Again, `hasSignedParts` detection is wrong, lower parts of unsigned ints may be encoded with negative values (because of the u32->i32 overflow handling), for example, `sdk.BigInteger([-1, 1])` corresponds to a valid `u64` value `8589934591`.
- `MAX_64` and `MAX_128` constants are wrong (especially, `MAX_128`, because it's in fact `MAX_28`). They are different for signed/unsigned values. If you intend to use them only as masks for int parts conversion, then it's ok. But they shouldn't be used for any boundary checks without explicit sign specified. I, personally, like hex notation for such masks. `0xffffffffffffffffffffffffffffffffn` looks much cleaner to me than `(1n << 128n) - 1n`, and it can be used directly, without precalculation and storing in the `const` variable. However, that's just personal preferences, not a hard requirement.
- It would be nice to have several standard methods to simplify logging and manipulation.
```js
valueOf() {return this.raw.valueOf()}
toString() {return this.raw.toString()}
toJSON() {return this.toString()}
```

**Comment by aristidesstaffieri** [comment]
Excited for something like this to land, for reference [this is an example of some code](https://github.com/stellar/soroban-react-payment/blob/main/src/utils/soroban.ts#L32) we tend to write in dapps or wallets to handle the contract call lifecycle. Ultimately it would be great to have a 1-to-1 decoder/encoder for every native type possible in Soroban from my perspective.

**Comment by sreuland** [comment]
wondering further out, how the other sdks will catch up at some point. On js, the discussion on https://github.com/stellar/js-xdr/pull/96 reached consensus to have this type of utility in stellar-base rather than lower in codegen(js-xdr/dts-xdr). The other sdk's don't have this extra module layer to leverage that separation pattern, just codegen from xdr-gen. 

seems like this presents an opportunity for consideration, If this utility were done at codegen time of js, then it may provide a precedent/pattern for codegen of other sdks to follow in xdr-gen(could js-xdr consolidate into xdr-gen..). Otherwise it looks like re-implementing the same utility in each sdk(java, go, python, etc). There is some precedent for adding functionality to generated code, looking at another tool like protobufs, which provides protoc plugin design, for the purpose of decorating the IDL generated code with more.

**Comment by Shaptic** [comment]
@sreuland I think you raise a valid point, and I'd love to see `js-xdr` + `dts-xdr` disappear into `xdrgen`, but that's a massive effort I doubt we'll see for quite some time. The impact on other SDKs in terms of the separation layer, though, is minimal I think:

The stellar-base vs. stellar-sdk divide is mostly superficial, nobody really uses stellar-base on its own, so this abstraction could just as feasibly have gone in the sdk. and the other SDKs don't _just_ do xdrgen -> code, they also have their own, language-specific abstraction layers. for example, the Go SDK has abstractions and helpers for the generated XDR built on top of it that are idiomatic to Go (e.g. [`txnbuild/asset.go`](https://github.com/stellar/go/blob/master/txnbuild/asset.go))

> provide a precedent/pattern for codegen of other sdks to follow in xdr-gen

But this codegen would be **very** specific to the other language, to the degree that it'd be the same amount of work as just writing that in the SDK itself. The implementation here is somewhat specific to JavaScript: the complexity of dealing with numbers (e.g. `Number` vs. `BigInt`) and bitwise operations (e.g. all bitwise operations like `<<` treating values as signed) in JavaScript is unique to this language. Other SDKs have different mechanisms (e.g. Python supports arbitrary length integers natively but Java uses `BigInteger`) and will need custom code to replicate this same abstraction layer

**Comment by Shaptic** [comment]
@orbitlens I was hoping nobody would peek too deeply into the code yet, just the interface :sweat_smile: but thank you for your feedback nonetheless! I came to some of the same conclusions yesterday when iterating. The interface still needs to evolve a bit and there are some obnoxious JavaScript quirks to overcome before I can determine whether or not it's even feasible (for example, converting a i64 to an i128 is surprisingly annoying because of how bitwise operations work in JS). 

Stay tuned for refinements!

**Comment by Shaptic** [comment]
Okay, I have some design & implementation updates:

 * The main interface change is that it only accepts a single value now rather than a list. This is because you can use the underlying XDR types (i.e. `xdr.LargeInt` child classes) if you want lower-level construction of these values values.

 * The implementation itself is mostly complete (draft PR: #620), but I need many more conversion tests before I can consider it to be ready for review, and I need to publish a corresponding PR to js-xdr.

Here's the new interface (in the form of a TypeScript type):

```ts
export type ScIntType = 'i32' | 'u32' | 'i64' | 'u64' | 'i128' | 'u128' | 'i256' | 'u256';
export class ScInt {
  constructor(value: number|bigint|string|ScInt, opts: { type: ScIntType });
  static fromScVal(scv: xdr.ScVal): bigint;

  toNumber(): number;
  toBigInt(): BigInt;

  toI64(): xdr.ScVal;
  toU64(): xdr.ScVal;
  toI128(): xdr.ScVal;
  toU128(): xdr.ScVal;
  toI256(): xdr.ScVal;
  toU256(): xdr.ScVal;

  valueOf(): any; // FIXME
  toString(): string;
  toJSON(): {
    value: string;
    type: ScIntType;
  }
}
```

and here's its usage (in the form of some examples):

```js
import sdk from "stellar-base";

// You have an ScVal from a contract and want to parse it into JS native.
const value = sdk.xdr.ScVal.fromXDR(someXdr, "base64");
const bigi = sdk.ScInt.fromScVal(value); // grab it as a BigInt
let sci = new ScInt(bigi);

sci.toNumber(); // gives native JS type (w/ size check)
sci.toBigInt(); // gives the native BigInt value
sci.toU64();    // gives ScValType-specific XDR constructs (with size checks)

// You have a number and want to shove it into a contract.
sci = sdk.ScInt(0xdeadcafebabe);
sci.toBigInt() // returns 244838016400062n
sci.toNumber() // throws: too large

// Pass any to e.g. a Contract.call(), conversion happens automatically
// regardless of the initial type.
const scValU128 = sci.toU128();
const scValI256 = sci.toI256();
const scValU64  = sci.toU64();

// Lots of ways to initialize:
sdk.ScInt("123456789123456789")
sdk.ScInt(123456789123456789n);
sdk.ScInt(1n << 140n);
sdk.ScInt(-42);
sdk.ScInt.fromScVal(scValU128); // from above

// If you know the type ahead of time (accessing `.raw` is faster than
// conversions), you can specify the type directly (otherwise, it's
// interpreted from the numbers you pass in):
const i = sdk.ScInt(123456789n, { type: "u256" });

// For example, you can use the underlying `sdk.U256` and convert it to an
// `xdr.ScVal` directly like so:
const scv = new xdr.ScVal.scvU256(i.raw);

// Or reinterpret it as a different type (size permitting):
const scv = i.toI64();
```

What do you guys think about this simplified interface? I think this alleviates the concerns that everyone had.

**Comment by Shaptic** [comment]
Some specific feedback responses:

@piyalbasu to your comment about "includ[ing] a .toU32 method to convert JS numbers to u32's", I think this is unnecessary because both creating and extracting them is easy (i.e. `new ScVal.scvU32(1).value() === 1`). However, you can still do _parsing_ with this abstraction via: `ScInt.fromScVal()` (just not creation)

@orbitlens @paulbellamy I dropped support for constructing from arrays of integers entirely, which simplifies handling signedness. you can still pass an optional `{ type }` to force it, and if you really need arrays, you can use the specific XDR wrapper types (i.e. `xdr.LargeInt` children)

**Comment by paulbellamy** [comment]
Do we need some `isU64()` type methods which do the size-checking and positive-check? Might be cleaner than a try/catch?

The second scenario (where you know the type) seems like it might be simpler if the ScVal union constructor took the variety of types directly, so you could do: `new xdr.ScVal.scvU256(val: number | bigint | ScInt)`. But either way is fine.

**Comment by paulbellamy** [comment]
I mean you wouldn't have to mess with the generated xdr too much. You could maybe do the `lodash/extend` type trick we used to do in js-xdr to override the generated method, with something like:

```ts
// int-mixin.ts
import xdr from '...';

function patch(method: string, toScVal: (i: sdk.ScInt) => xdr.ScVal) {
  const wrap = xdr.ScVal[method]

  xdr.ScVal[method] = (val: number|bigint|string|ScInt) =>
    wrap(toScVal(sdk.ScInt(val)));
}

patch('scvU256', xdr.ScInt.prototype.toU256);
// ... patch the others
```

**Comment by orbitlens** [comment]
@Shaptic Great work! But I think that the proposed interface is still misleading. 

Usage of the constructor with the semantics `constructor(value: number|bigint|string|ScInt, opts: { type: ScIntType });` may result in unwanted side effects when dealing with arguments other than `ScInt`. It's impossible to automatically determine underlying int size from number/bigint/string types. So people will have to **always** provide the second optional argument to explicitly specify the type, otherwise the code will be simply unsafe to use.

How about this definition?
```ts
constructor(value: string|number|bigint|bigint[], type: ScIntType);
constructor(scIntValue: ScInt);
```

This way, if a developer provides an `ScInt` value, the second argument is ignored, and the type is determined from the `ScInt` itself. With any other input type, the constructor should set the int size based on the second argument. And this way we still retain the ability to initialize from a bigint array. For those of us who code in JS rather than TS, it's much more convenient to work with "wildcard" functions that can handle all types of arguments.

I really don't like the internal `nearestBigIntSize` function -- it's a potential performance bottleneck. Code on lines [157-182 ](https://github.com/stellar/js-stellar-base/pull/620/files#diff-49d3f9fbd2e0287b5c41b97d159e41e325610709ac8461c3451b65306cdbe14cR157-R182) is excessive to my mind.

**Comment by Shaptic** [comment]
@orbitlens thanks for your comments! I guess this whole discussion about arrays & types needs a different question answered first for me to understand your perspective better: What are the expected use cases of this abstraction? In my mind, you have these:

* If you want a specific `ScVal` type, you can either specify the `{ type }` as the second parameter OR call the appropriate `.toU128()` (or whichever) and it'll give you the correct type encoded regardless of your input value.
* If you want to build an XDR type from an array of integers, you can use the `LargeInt` types directly (e.g. `new U128([...values...])`.
  - If you _still_ want the new abstraction after this (because you want to interact with contracts or whatever), I just added `fromXdrInt(i: xdr.LargeInt): ScInt` to cover that case.
* If you don't care and just want the smallest XDR type that fits your value, that's the default behavior. And sure, it's a little slower because of type detection, but that's exactly the point of "I don't care what XDR the abstraction uses."

So, in my mind, all of the cases are covered. What am I missing here?

**Comment by orbitlens** [comment]
@Shaptic 
> If you don't care and just want the smallest XDR type that fits your value, that's the default behavior.

I cannot think of any use-case where I might need XDR with random type. What I'm trying to say, it's a dangerous behavior, because you never know in advance what you will get in the output, of course, except the situation when you know exactly all the inputs in advance. But in this case, one can just use constants mapping instead. 

If you are ok to accept random values, than you'll almost certainly get an error when you try using the constructed value if it's required for anything else except the toString() conversion or testing purpose. Soroban contracts require strong typing, low-level XDR entries require strong typing, every XDR-related piece of code I ever created required strong typing.

For example, I can parse the response of invocation for Soroban contract with this and output to console, but I won't be able to correctly assemble the invocation arguments for a Soroban contract function without explicitly specifying the type. That's the real world example from our recent hands-on experience.

From my point of view, `type` is always **mandatory** for any sane developer writing anything more complex than `console.log(new ScInt('1'))`. And if we agree that `type` can't be an optional parameter, I don't see a reason why we can't allow `bigint[]` as argument -- the underlying XDR type will automatically parse the arguments, no need to write any additional code in the wrapper class.

**Comment by Shaptic** [comment]
> you'll almost certainly get an error when you try using the constructed value

But there's nothing you **can** use the constructed value for without specifying the type! It's not an XDR type itself, so why does it matter what the `ScInt` instance uses _internally_ to represent the input? "Dangerous" feels a little excessive since you will always need to call `.toU128()` or its equivalent anyway before you actually use it: that's exactly where you get the strong typing you're looking for. Making the type mandatory just makes things mildly redundant, e.g. you propose

```ts
const sci = new ScInt('u128', someInputField.value);
someContract.call('add', sci.toU128(), sci.toI64()); // or
sci.toU128().u128(); // to get the xdr.Uint128Parts
```

vs. the existing

```diff
-const sci = new ScInt('u128', someInputField.value);
+const sci = new ScInt(someInputField.value);
```

I _guess_ you could argue that `sci.toU128()` becomes marginally faster in the strongly typed case, but that feels like a micro-optimization?

**Comment by orbitlens** [comment]
It’s not a hill I’m willing to die on. Just expressed my thoughts on this matter. I have no other arguments except those I already outlined. Please proceed with any interface you see fit. In this case `ScInt` will be useless for us, so we'll be working with `LargeInt` types directly, no problems.

We just need a simple way to encode/decode `i128` and `i256` in our applications, that's a blocker for us. The sooner this functionality is released -- the better.

**Comment by paulbellamy** [comment]
@orbitlens 
Could you elaborate on this?
> It's impossible to automatically determine underlying int size from number/bigint/string types. So people will have to always provide the second optional argument to explicitly specify the type, otherwise the code will be simply unsafe to use.

Afaik, the LargeInt internal representation is a bigint, which should be able to contain any size of value. The constructor would never fail, so I don't see how it is unsafe.

The output conversion methods might fail (e.g. `.toU32()`), but that would always be the case.

(This is assuming there is no the `bigint[]` constructor input type, because any construction-from-bytes does require an explicit typing)

**Comment by Shaptic** [comment]
I realized this morning that this should be really easy to solve: 30fb037 adds an `XdrInt` which becomes the base class for `ScInt`, except it requires a type parameter and accepts arrays. I think this is a perfect compromise, @orbitlens?

**Comment by orbitlens** [comment]
> I realized this morning that this should be really easy to solve: [30fb037](https://github.com/stellar/js-stellar-base/commit/30fb0371d7ca54ce96504120ab1f2c54d0487d8b) adds an `XdrInt` which becomes the base class for `ScInt`, except it requires a type parameter and accepts arrays. I think this is a perfect compromise, @orbitlens?

Thanks, @Shaptic ! LGTM.
The only suggestion [here](https://github.com/stellar/js-stellar-base/commit/30fb0371d7ca54ce96504120ab1f2c54d0487d8b#diff-49d3f9fbd2e0287b5c41b97d159e41e325610709ac8461c3451b65306cdbe14cR83): 
```js
values = values.map(i => {
    if (typeof i === 'bigint')
        return i;
    if (i instanceof XdrInt)
        return i.toBigInt();        
    return BigInt(i);
});
```
I suspect that people will mostly provide `bigint` as an argument, so this microoptimization should be slightly faster.

**Comment by silence48** [comment]
> The stellar-base vs. stellar-sdk divide is mostly superficial, nobody really uses stellar-base on its own

I'm just gonna point out that I often use stellar base because I work heavily with CF workers and pages., and you can't use axios nor eventsource there.

**Comment by Shaptic** [comment]
Closed by #620! It will be released in the next release of stellar-base@soroban.

**Comment by orbitlens** [comment]
@Shaptic are there any chances to see `stellar-base` and `stellar-base@soroban` packages merged into a single package in the nearest future? Currently we have to include both dependencies to `package.json` of our applications which complicates the code and results in additional bugs.

**Comment by Shaptic** [comment]
@orbitlens that's definitely a goal, but because there are massive breaking XDR changes (Soroban and otherwise), the "mainline" version is necessary because of the "curr" vs. "next" protocol division, so anyone wanting to build on stellar today cannot use the `@soroban` tag
